refactor: clarify public API

ajbozarth opened the pull request:
### Which issue(s) does this pull-request address?

Fixes #547

### Description

This clarifies out public API by accomplishing the following:
- Go through and update all current code to match the public/protected/private/readonly status that exists in TS
- Update all the project `__init__.py` file's `__all__` lists to only export classes that should be publicly available according to the TS implementation and our current examples.
- Updating all our import paths to be more specific in our internal code, and more general in our examples and tests.
   - No internal imports should import from root

To Do:
- [x] Decide what classes should be available to import at root (ie `from beeai_framework import X`) rather than more specific paths. If taken as is this PR will leave no classes available at root (thus be a breaking change)
  

### Checklist

#### General
- [x] I have read the appropriate contributor guide: [Python](https://github.com/i-am-bee/beeai-framework/blob/main/CONTRIBUTING.md)
/ [TypeScript](https://github.com/i-am-bee/beeai-framework/blob/main/CONTRIBUTING.md)
- [x] I have signed off on my commit: [Python instructions](https://github.com/i-am-bee/beeai-framework/blob/main/python/CONTRIBUTING.md#developer-certificate-of-origin-dco) / [TypeScript instructions](https://github.com/i-am-bee/beeai-framework/blob/main/typescript/CONTRIBUTING.md#developer-certificate-of-origin-dco)
- [x] Commit messages and PR title follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary)
- [x] Appropriate label(s) added to PR: `Python` for Python changes, `TypeScript` for TypeScript changes

#### Code quality checks
- [x] Linting passes: Python `poe lint` or `poe lint --fix` / TypeScript `yarn lint` or `yarn lint:fix`
- [x] Formatting is applied: Python `poe format` or `poe format --fix` / TypeScript: `yarn format` or `yarn format:fix`
- [x] Static type checks pass: Python `poe type-check`

#### Testing
- [x] Unit tests pass: Python `poe test --type unit` / TypeScript `yarn test:unit`
- [x] E2E tests pass: Python `poe test --type e2e` / TypeScript: `yarn test:e2e`
- [x] Integration tests pass: Python `poe test --type integration`
- [x] Tests are included (for bug fixes or new features)

#### Documentation
- [x] Documentation is updated
- [x] Embedme embeds code examples in docs. To update after edits, run: Python `poe docs --type build`

ajbozarth commented:
I know this is a huge PR with changes to close to every file, so in order to aid in review I made a separate commit for each submodule, then a final commit for cleaning up the imports in examples and tests, and any I missed along the way.

Tomas2D reviewed (changes requested):
I like it, although you added a lot of unnecessary getters/setters. 
In addition, I think that the modules' `__init__.py` should not export everything from all nested submodules. Like `tools/__init__.py` should not export all tools but rather the base tool class + `tool` decorator + types.

ajbozarth commented on python/beeai_framework/adapters/__init__.py:
```diff
@@ -12,4 +12,29 @@
+from beeai_framework.adapters.watsonx.backend.chat import WatsonxChatModel
+from beeai_framework.adapters.xai.backend.chat import XAIChatModel
 
+__all__ = [
+    "AmazonBedrockChatModel",
+    "AnthropicChatModel",
+    "AzureOpenAIChatModel",
+    "GroqChatModel",
+    "LangChainTool",
+    "LangChainToolRunOptions",
+    "LiteLLMChatModel",
+    "OllamaChatModel",
+    "OpenAIChatModel",
+    "VertexAIChatModel",
+    "WatsonxChatModel",
+    "XAIChatModel",
+]
```
That's fair, this was my one attempt at streamlining >.<

ajbozarth commented on python/beeai_framework/agents/react/agent.py:
```diff
@@ -68,13 +68,13 @@ def __init__(
     ) -> None:
         super().__init__()
-        self.input = ReActAgentInput(
+        self._input = ReActAgentInput(
```
This PR will already be a `BREAKING CHANGE` due to the change in export locations (no more root exports).

ajbozarth commented on python/beeai_framework/cancellation.py:
```diff
@@ -110,3 +110,6 @@ def abort_handler() -> None:
             signal.remove_event_listener(abort_handler)
+
+
+__all__ = ["AbortSignal"]
```
moved and updated to export from utils

ajbozarth commented on python/beeai_framework/tools/code/sandbox.py:
```diff
@@ -46,7 +46,7 @@ class SandboxToolOptions(ToolRunOptions):
     def __init__(self, options: SandboxToolOptions) -> None:
         super().__init__()
-        self._options = options
+        self.__options = options
```
`_options` now exists on the parent class and it was complaining, I just renamed it `_tool_options` instead of `__options`

ajbozarth commented:
I've made a bunch of edits to address the review comments and minimize unnecessary code changes. The remaining discussion topics for discussion are:
1. How we should handle "private" properties in `Message`, `RunContext`, `Emitter`, and `ReadOnlyMemory` (as specifically called out by @Tomas2D )
2. should some classes be exported from within their own file rather than in `__init__.py` (@Tomas2D called out `MCPTool` and `AgentWorkflow` specifically )

ajbozarth commented on python/beeai_framework/workflows/__init__.py:
```diff
@@ -12,8 +12,29 @@
 
-__all__ = ["Workflow", "WorkflowError", "WorkflowReservedStepName"]
+__all__ = [
+    "AgentWorkflow",
+    "AgentWorkflowInput",
```
moved to subdirectory as discussed

ajbozarth commented on python/docs/tutorials.md:
```diff
@@ -78,15 +78,14 @@ from dotenv import load_dotenv
-from beeai_framework.tools.mcp_tools import MCPTool
-from beeai_framework.tools.weather.openmeteo import OpenMeteoTool
+from beeai_framework.memory import TokenMemory
+from beeai_framework.tools import MCPTool
```
moved to subdirectory as discussed

ajbozarth commented:
For reference, adding this here. I'll include it in the final commit message:

```
BREAKING CHANGES:
- Removed ability to import classes from beeai_framework without a path
- Moved AbortSignal from beeai_framework.cancellation to beeai_framework.utils
- Moved MCPTool from beeai_framework.tools.mcp_tools to beeai_framework.tools.mcp
```

Tomas2D reviewed (approved):
👍🏻
